### ISSUE TITLE
Need Ability to Add Additional Properties to Links

### ISSUE THREAD

**Issue by EarthCitizen** [spring-projects/spring-hateoas#1698]
Our internal standard requires the HTTP method on all links. This does not currently appear to be possible. I could add affordances, but, as far as I can tell, this requires adding them to a "root" link, which would not have the HTTP method. And there might not always be a self link.

The only workaround, that I am currently aware of, is to subclass `Link` and add the HTTP method field, and manually create these after the regular links are created with a builder.

It would appear that the workaround above will no longer work when the deprecated public constructors of `Link` are removed, or if the class is made final. At that point we would seem to be forced to either use a brute-force methodology like with reflection utilities or something like ByteBuddy as there are no setters on the underlying properties of `Link`. Or we would not be able to use Spring HATEOAS at all, which would be unfortunate for an issue so small.

What would be great is if the `Link` class had a way to add ad-hoc properties.

**Comment by Ickbinet** [comment]
Well... that "method attribute on links" question is long debated and the answer is always (mostly) : NO. Do not ask me why...

I am also not sure if the link spec (https://datatracker.ietf.org/doc/html/rfc5988) allows it. I guess they always had link =navigation=get method in mind.

Imho the affordance is the solution. You can add an affordance to any link, not only root. 
And yes, this does not solve the problem for clients which do not understand/use affordances...

**Comment by EarthCitizen** [comment]
@Ickbinet I could be misunderstanding, but the only way to add affordances is to have an existing `Link`, which its affordances will have the HTTP method, but that link itself will not.

**Comment by odrotbohm** [comment]
Could be that I am missing something, but any custom serialization (via Mixin or even Serializer) registered for the `Link` created via the `WebMvcLinkBuilder` has access to its default `Affordance` that contains the HTTP method associated with it and can be included in the serialization.

**Comment by EarthCitizen** [comment]
@odrotbohm I was not aware that the `WebMvcLinkBuilder` added affordances automatically. How is the default identified? Is it simply the first affordance in the list?

**Comment by EarthCitizen** [comment]
As a note to anyone not using the built-in media types (i.g. `application/json`), it is necessary to do the following:
```java
public class JsonAffordanceModel extends AffordanceModel {
    public JsonAffordanceModel(ConfiguredAffordance configured) {
        super(configured.getNameOrDefault(), configured.getTarget(), configured.getMethod(), configured.getInputMetadata(),
            configured.getQueryParameters(), configured.getOutputMetadata());
    }
}
```
```java
public class JsonAffordanceModelFactory implements AffordanceModelFactory {
    @Override
    public MediaType getMediaType() {
        return MediaType.APPLICATION_JSON;
    }

    @Override
    public AffordanceModel getAffordanceModel(ConfiguredAffordance configured) {
        return new JsonAffordanceModel(configured);
    }
}
```
And in `resources/META-INF/spring.factories`:
```
org.springframework.hateoas.mediatype.AffordanceModelFactory=\
 some.package.JsonAffordanceModelFactory
```

**Comment by odrotbohm** [comment]
> @odrotbohm I was not aware that the `WebMvcLinkBuilder` added an affordance automatically for the link. How is the default identified? Is it simply the first affordance in the list?

You get an affordance for every `HttpMethod` supported on the mapping. Whichever you declare first will be the first. Effectively, you'll have to process all of them as a controller method, and thus a link pointing to it, can support multiple HTTP methods simultaneously.

> As a note to anyone not using the built-in media types (e.g. application/json), it is necessary to do the following…

That's correct. I'll extend the section of the reference docs dealing with custom media type registration to include that information.

**Comment by EarthCitizen** [comment]
@odrotbohm So what I did was create a custom utility class that wraps around `WebMvcLinkBuilder` and adds the correct affordances and relationships based on Spring annotations. Then I needed to create a custom serializer for `Links` that looks for affordance HTTP methods and renderts the links according to our internal standard. Therefore, I do not need the original feature I was requesting because I found a workaround based on your advice. Closing this.
